solr/solrj/src/java/org/apache/solr/common/util/CompressionUtil.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.common.util;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.zip.DataFormatException;
+import java.util.zip.Deflater;
+import java.util.zip.DeflaterInputStream;
+import java.util.zip.Inflater;
+
+/**
+ * Utility class for dealing with LZ4 compressed data, such as state.json in Zookeeper if configured to be compressed
+ */
+public class CompressionUtil {
+
+    /**
+     * Decompresses zlib compressed bytes, returning the uncompressed data as a byte[]
+     * @param data the input zlib compressed data to decompress
+     * @return the decompressed bytes
+     * @throws DataFormatException - The data is not zlib compressed data
+     */
+    public static byte[] decompressCompressedBytes(byte[] data) throws DataFormatException {
+        if (data == null) return null;
+        Inflater inflater = new Inflater();
+        inflater.setInput(data);
+        byte[] buf = new byte[100];
+        int read;
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        while (!inflater.finished()) {
+            read = inflater.inflate(buf, 0, buf.length);
+            os.write(buf, 0, read);
+        }
+        return os.toByteArray();
+    }
+
+    /**
+     * Compresses bytes into zlib compressed bytes using java native compression
+     * @param data the input uncompressed data to be compressed
+     * @return zlib compressed bytes
+     * @throws IOException
+     */
+    public static byte[] compressBytes(byte[] data) throws IOException {
+        DeflaterInputStream deflaterInputStream = new DeflaterInputStream(new ByteArrayInputStream(data), new Deflater(Deflater.BEST_SPEED));
+        byte[] buf = new byte[100];
+        int read;
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
```

Comment:
analogous to above -- we can probably make a reasonable initial size estimate based on expected compression ratio. If we over-estimate by a bit, we should be able to avoid unnecessary arraycopies while growing the output buffer.

Reply:
Also, in `compressBytes()` we can work directly with `Deflater`, avoiding the need for the wrapping `DeflaterInputStream`, `ByteArrayInputStream`, etc. Just need to be careful to call `Deflater.end()` before allowing one-off Deflaters to go out of scope (mentioning because it's an easy thing to overlook -- in fact since the current code specifies a non-default Deflater, even sticking with the _current_ code could I think [introduce a memory leak](https://stackoverflow.com/a/13550595) unless we keep a ref to the specified Deflater and call `end()` on it when it goes out of scope!). -- the nowrap arg can be used safely I think, i.e.: `new Deflater(Deflater.BEST_SPEED, true)`.

Also (this is not as important): if each ZkStateWriter is only accessed by a single thread (which I think is the case, in Overseer), it should be possible to pass in our own Deflater to the compress method and do `deflater.reset()` for each time it's used. Naive testing suggests that Deflater ctor initialization is surprisingly heavy, and `reset()` should be not-insignificantly faster, if we're certain it can be done safely ... this may be getting to cute though.